### PR TITLE
Putting spark-hdfs on BackendModule

### PR DESCRIPTION
**Pull request by rabbitonweb**
Part of #2216

### CODE REVIEW

**Comment by alissapajer** [comment]
@rabbitonweb is this ready for review? If so, can you assign someone? 

And also it needs a semantic version label.

**Comment by rabbitonweb** [comment]
@alissapajer not yet ready for review, will be soon.

**Comment by rabbitonweb** [comment]
@wemrysi this is ready for review, `ChrootedInterpreter` added as discussed :)

File: connector/src/main/scala/quasar/connector/ChrootedInterpreter.scala [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+import slamdata.Predef._
+import quasar.contrib.pathy._
+import quasar.fp.free._
+import quasar.fs._, mount.BackendDef.DefErrT
+
+import pathy.Path._
+import scalaz._, Scalaz._, concurrent.Task
+
+trait ChrootedInterpreter extends BackendModule {
```
Comment (wemrysi) [inline connector/src/main/scala/quasar/connector/ChrootedInterpreter.scala]: Mind changing this to be a `BackendModule` mixin (via `self: BackendModule`) similar to the `Managed*` mixins? I think it clarifies that the intent of the `trait` is to augment `BackendModule` and isn't intended to be referenced directly.

Comment (rabbitonweb) [reply connector/src/main/scala/quasar/connector/ChrootedInterpreter.scala]: but that does not allow overriding of methods, right?

File: connector/src/main/scala/quasar/connector/ChrootedInterpreter.scala [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+  def prefix(cfg: Config): ADir
```
[the remark this answers is not in the document]
Comment (rabbitonweb) [reply connector/src/main/scala/quasar/connector/ChrootedInterpreter.scala]: 👍

File: connector/src/main/scala/quasar/connector/ChrootedInterpreter.scala (hunk 1 above)
Comment (wemrysi) [reply connector/src/main/scala/quasar/connector/ChrootedInterpreter.scala]: Not sure, if not then I guess we don't have a choice.

File: sparkcore/src/main/scala/quasar/physical/sparkcore/fs/SparkCoreBackendModule.scala [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+package quasar.physical.sparkcore.fs
+
+import slamdata.Predef._
+import quasar._
+import quasar.connector.BackendModule
+import quasar.contrib.pathy._
+import quasar.common._
+import quasar.fp._, free._
+import quasar.fs._, FileSystemError._//, PathError._
```
[the remark this answers is not in the document]
Comment (rabbitonweb) [reply sparkcore/src/main/scala/quasar/physical/sparkcore/fs/SparkCoreBackendModule.scala]: 👍

File: sparkcore/src/main/scala/quasar/physical/sparkcore/fs/local/SparkLocalBackendModule.scala [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+import org.apache.spark.rdd._
+import pathy.Path._
+import scalaz.{Failure => _, _}, Scalaz._
+import scalaz.concurrent.Task
+
+final case class LocalConfig(sparkConf: SparkConf, prefix: ADir)
+
+object SparkLocalBackendModule extends SparkCoreBackendModule {
+
+  override def interpreter(cfg: Config): DefErrT[Task, (BackendEffect ~> Task, Task[Unit])] = {
```
[the remark this answers is not in the document]
Comment (rabbitonweb) [reply sparkcore/src/main/scala/quasar/physical/sparkcore/fs/local/SparkLocalBackendModule.scala]: LOL 👍

File: sparkcore/src/main/scala/quasar/physical/sparkcore/fs/hdfs/SparkHdfsBackendModule.scala [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+  } yield {
+    sc.addJar(posixCodec.printPath(jar))
+    sc
+  }
+
+  private def toPath(apath: APath): Free[Eff, Path] = lift(Task.delay {
+    new Path(posixCodec.unsafePrintPath(apath))
+  }).into[Eff]
+
+  object HdfsWriteFileModule extends WriteFileModule {
```
[the remark this answers is not in the document]
Comment (rabbitonweb) [reply sparkcore/src/main/scala/quasar/physical/sparkcore/fs/hdfs/SparkHdfsBackendModule.scala]: https://github.com/quasar-analytics/quasar/issues/2722

File: sparkcore/src/main/scala/quasar/physical/sparkcore/fs/SparkCoreBackendModule.scala [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+import org.apache.spark.SparkContext
+import pathy.Path._
+import matryoshka._
+import matryoshka.implicits._
+import scalaz.{Failure => _, _}, Scalaz._
+import scalaz.concurrent.Task
+
+trait SparkCoreBackendModule extends BackendModule {
+
+  // conntector specificc
```
[the remark this answers is not in the document]
Comment (rabbitonweb) [reply sparkcore/src/main/scala/quasar/physical/sparkcore/fs/SparkCoreBackendModule.scala]: 👍

File: connector/src/main/scala/quasar/connector/BackendModule.scala [hunk 7]
```diff
@@ -55,8 +55,16 @@ trait BackendModule {
 
   implicit class LiftBackend[A](m: M[A]) {
     val liftB: Backend[A] = m.liftM[ConfiguredT].liftM[PhaseResultT].liftM[FileSystemErrT]
+    def withLog(pr: PhaseResult): Backend[A] =
```
[the remark this answers is not in the document]
Comment (rabbitonweb) [reply connector/src/main/scala/quasar/connector/BackendModule.scala]: I've used `  final def :++>(w: ⇒ W)(implicit B: Bind[F]): F[A] `

**Comment by rabbitonweb** [comment]
@wemrysi ready

**Pull request merged**